go/arrow/util/protobuf_reflect.go

```diff
@@ -0,0 +1,447 @@
+	}
+	return ProtobufFieldReflection{
+		fd,
+		psr.message.Get(fd),
+		fv,
+		psr.SchemaOptions,
+	}
+}
+
+type ProtobufFieldReflection struct {
+	descriptor protoreflect.FieldDescriptor
+	prValue    protoreflect.Value
+	rValue     reflect.Value
+	SchemaOptions
+}
+
+func (pfr ProtobufFieldReflection) isStruct() bool {
+	return pfr.descriptor.Kind() == protoreflect.MessageKind && !pfr.descriptor.IsMap() && pfr.rValue.Kind() != reflect.Slice
+}
+
+func (pfr ProtobufFieldReflection) isMap() bool {
+	return pfr.descriptor.Kind() == protoreflect.MessageKind && pfr.descriptor.IsMap()
+}
+
+func (pfr ProtobufFieldReflection) isList() bool {
+	return pfr.descriptor.IsList() && pfr.rValue.Kind() == reflect.Slice
+}
+
+func (pfr ProtobufFieldReflection) getListLength() int {
+	return pfr.prValue.List().Len()
+}
+
+func (pfr ProtobufFieldReflection) getMapLength() int {
+	return pfr.prValue.Map().Len()
+}
+
+func (plr ProtobufListReflection) generateListItems() chan ProtobufFieldReflection {
+	out := make(chan ProtobufFieldReflection)
+
+	go func() {
+		defer close(out)
+		for i := 0; i < plr.prValue.List().Len(); i++ {
+			out <- ProtobufFieldReflection{
+				descriptor:    plr.descriptor,
+				prValue:       plr.prValue.List().Get(i),
+				rValue:        plr.rValue.Index(i),
+				SchemaOptions: plr.SchemaOptions,
+			}
+		}
+	}()
+
+	return out
+}
+
+func (pfr ProtobufFieldReflection) getDataType() arrow.DataType {
+	var dt arrow.DataType
+
+	typeMap := map[protoreflect.Kind]arrow.DataType{
+		//Numeric
+		protoreflect.Int32Kind:    arrow.PrimitiveTypes.Int32,
+		protoreflect.Int64Kind:    arrow.PrimitiveTypes.Int64,
+		protoreflect.Sint32Kind:   arrow.PrimitiveTypes.Int32,
+		protoreflect.Sint64Kind:   arrow.PrimitiveTypes.Int64,
+		protoreflect.Uint32Kind:   arrow.PrimitiveTypes.Uint32,
+		protoreflect.Uint64Kind:   arrow.PrimitiveTypes.Uint64,
+		protoreflect.Fixed32Kind:  arrow.PrimitiveTypes.Uint32,
+		protoreflect.Fixed64Kind:  arrow.PrimitiveTypes.Uint64,
+		protoreflect.Sfixed32Kind: arrow.PrimitiveTypes.Int32,
+		protoreflect.Sfixed64Kind: arrow.PrimitiveTypes.Int64,
+		protoreflect.FloatKind:    arrow.PrimitiveTypes.Float32,
+		protoreflect.DoubleKind:   arrow.PrimitiveTypes.Float64,
+		//Binary
+		protoreflect.StringKind: arrow.BinaryTypes.String,
+		protoreflect.BytesKind:  arrow.BinaryTypes.Binary,
+		//Fixed Width
+		protoreflect.BoolKind: arrow.FixedWidthTypes.Boolean,
+		// Enum
+		protoreflect.EnumKind: arrow.PrimitiveTypes.Int32,
+		// Struct
+		protoreflect.MessageKind: nil,
+	}
+	dt = typeMap[pfr.descriptor.Kind()]
+
+	if pfr.isStruct() {
+		dt = pfr.AsStruct().getDataType()
+	}
+
+	if pfr.isMap() {
+		dt = pfr.AsMap().getDataType()
+	}
+
+	if pfr.isList() {
+		dt = pfr.AsList().getDataType()
+	}
+	return dt
+}
+
+func getBuilders(s *arrow.Schema, m memory.Allocator) []array.Builder {
+	var builders []array.Builder
+
+	for _, f := range s.Fields() {
+		builders = append(builders, array.NewBuilder(m, f.Type))
+	}
+	return builders
```

Comment:
it would probably be more efficient to create a RecordBuilder and return that for this